pkg/scheduler/api/namespace_info.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	// QuotaKey is the key in ResourceQuota.spec.hard indicating the weight of thi namespace
+	QuotaKey = "volcano.sh/namespace.weight"
+	// DefaultWeight is the default weight of namespace
+	DefaultWeight = 1
+)
+
+// NamespaceInfo records information of namespace
+type NamespaceInfo struct {
+	Name string
+	// Weight is the highest weight among many ResourceQuota.
+	Weight int64
+}
+
+// GetWeight returns weight of a namespace, any invalid case would get default value
+func (n *NamespaceInfo) GetWeight() int64 {
+	if n == nil {
+		return DefaultWeight
+	}
+	if n.Weight == 0 {
+		return DefaultWeight
+	}
+	return n.Weight
+}
+
+type quotaItem struct {
+	name   string
+	weight int64
+}
+
+func quotaItemKeyFunc(obj interface{}) (string, error) {
+	item, ok := obj.(*quotaItem)
+	if !ok {
+		return "", fmt.Errorf("obj with type %T could not parse", obj)
+	}
+	return item.name, nil
+}
+
+// for big root heap
+func quotaItemLessFunc(a interface{}, b interface{}) bool {
+	A := a.(*quotaItem)
+	B := b.(*quotaItem)
+	return A.weight > B.weight
+}
+
+// NamespaceCollection will record all details about namespace
+type NamespaceCollection struct {
+	Name string
+
+	weightMu    sync.Mutex
+	quotaWeight *cache.Heap
```

Comment:
To my understanding, only need to keep the biggest weight.

Even if you want all the weights, there is a `PriorityQueue` in volcano, does it satisfy your requirement?

Reply:
The only biggest weight will be kept in `NamespaceInfo` after `cache.Snapshot()`.

At here, `NamespaceCollection` store all the information about this namespace and will be able to handle the weight changes in `Quota` resource (such as try to decrease the biggest weight).

I'll have a look at `PriorityQueue` later~

Reply:
The `PriorityQueue` is a simple implementation of `container/heap`, but it could not support the `Update` action, and delete some item in heap directly is also not supported.

So, It it also the reason why I choose the Heap in k8s cache before.

Reply:
IC,  this is to support resourcequota update